src/math/fft_base.h

```diff
@@ -0,0 +1,190 @@
+#pragma once
+
+#include <vector>
+#include "../image_exception.h"
+
+namespace FFT 
+{
+    // This class template is the base for storing complex-valued data ([real, imaginary]) for
+    // the purpose of using in a Fast Fourier Transform. The template parameter is the type of 
+    // data to use. Instances of this class template are meant to be inherited by a sub-class.
+    // The sub-class should define _allocateData(), _freeData(), and _copyData(). 
+    //
+    // The function template definitions are in fft_base.tpp, and this file is included at the 
+    // end of this file (after the class template declarations).
+
+    template <typename DataType>
+    class BaseComplexData 
+    {
+    public:
+        BaseComplexData();
+
+        BaseComplexData<DataType> & operator=( const BaseComplexData<DataType> & data );
+        BaseComplexData<DataType> & operator=( BaseComplexData<DataType> && data );
+
+        void resize( uint32_t width_, uint32_t height_ );
+
+        DataType * data(); // returns a pointer to data
+        const DataType * data() const;
+        bool empty() const; // returns true is array is empty (unallocated)
+        uint32_t width() const
+        {
+            return _width;
+        }
+        uint32_t height() const
+        {
+            return _height;
+        }
+        bool dimensionsMatch( const BaseComplexData<DataType> & other) const;
+        bool dimensionsMatch( uint32_t width, uint32_t height) const;
+
+    protected:
+        uint32_t _width, _height;  
+        DataType * _data;
+
+        virtual void _allocateData(size_t nBytes) = 0; // Override to deal with actual memory allocation.
```

Comment:
I thought that the variable measures a size is covered by the typename `size`. I thought it would be nice to have a name to indicate what units the size is in, i.e. number of bytes vs number of array elements. Would the name `bytes` also be good?

Reply:
I will go with your suggestion of `size`.

Done (technically you made the change, thank you).